src/commands/mod.rs

```diff
@@ -1009,6 +1032,229 @@ impl DaemonControl {
+        } else {
+            HashMap::new()
+        };
+
+        // If there was no previous change address, we set the change address for the replacement
+        // to our next change address. This way, we won't increment the change index with each attempt
+        // at creating the replacement PSBT below.
+        let change_address = prev_change_address.cloned().unwrap_or_else(|| {
+            let index = db_conn.change_index();
+            let desc = self
+                .config
+                .main_descriptor
+                .change_descriptor()
+                .derive(index, &self.secp);
+            desc.address(self.config.bitcoin_config.network)
+        });
+        // If `!is_cancel`, we take the previous coins as mandatory candidates and add confirmed coins as optional.
+        // Otherwise, we take the previous coins as optional candidates and let coin selection find the
+        // best solution that includes at least one of these. If there are insufficient funds to create the replacement
+        // transaction in this way, then we set candidates in the same way as for the `!is_cancel` case.
+        let mut candidate_coins: Vec<CandidateCoin> = prev_coins
+            .values()
+            .map(|c| CandidateCoin {
+                coin: *c,
+                must_select: !is_cancel,
+            })
+            .collect();
+        let confirmed_cands: Vec<CandidateCoin> = db_conn
+            .coins(&[CoinStatus::Confirmed], &[])
+            .into_values()
+            .filter_map(|c| {
+                // In case the user attempts RBF before the previous coins have been updated in the DB,
+                // they would be returned as confirmed and not spending.
+                if !prev_coins.contains_key(&c.outpoint) {
+                    Some(CandidateCoin {
+                        coin: c,
+                        must_select: false,
+                    })
+                } else {
+                    None
+                }
+            })
+            .collect();
+        if !is_cancel {
+            candidate_coins.extend(&confirmed_cands);
+        }
+        let max_sat_weight: u64 = self
+            .config
+            .main_descriptor
+            .max_sat_weight()
+            .try_into()
+            .expect("it must fit");
+        // Try with increasing feerate until fee paid by replacement transaction is high enough.
+        // Replacement fee must be at least:
+        // sum of fees paid by original transactions + incremental feerate * replacement size.
+        // Loop will continue until either we find a suitable replacement or we have insufficient funds.
+        let mut rbf_vsize = 0_u64;
+        loop {
+            // Note that `rbf_vsize` strictly increases on each iteration as otherwise `min_fee`
+            // will ensure that the PSBT meets the required replacement fee and the loop will exit.
+            let min_fee = mempool_entry.fees.descendant.to_sat() + rbf_vsize;
```

Comment:
Actually `rbf_vsize` does not strictly increase, as you re-set it to the same value on every iteration below. We should have a test which exercises the case where there is >1 iteration, this would have caught the infinite loop here.

Suggestion for a fix:
```diff
diff --git a/src/commands/mod.rs b/src/commands/mod.rs
index c662300..d6ee1d9 100644
--- a/src/commands/mod.rs
+++ b/src/commands/mod.rs
@@ -1021,11 +1021,12 @@ impl DaemonControl {
         // Replacement fee must be at least:
         // sum of fees paid by original transactions + incremental feerate * replacement size.
         // Loop will continue until either we find a suitable replacement or we have insufficient funds.
-        let mut rbf_vsize = 0_u64;
-        loop {
+        let mut replacement_vsize = 0;
+        for incremental_feerate in 0.. {
             // Note that `rbf_vsize` strictly increases on each iteration as otherwise `min_fee`
             // will ensure that the PSBT meets the required replacement fee and the loop will exit.
-            let min_fee = mempool_entry.fees.descendant.to_sat() + rbf_vsize;
+            let min_fee =
+                mempool_entry.fees.descendant.to_sat() + replacement_vsize * incremental_feerate;
             let rbf_psbt = match self.create_spend_internal(
                 &destinations,
                 &candidate_coins,
@@ -1050,17 +1051,20 @@ impl DaemonControl {
                     return Err(e);
                 }
             };
-            rbf_vsize = unsigned_tx_max_vbytes(&rbf_psbt.psbt.unsigned_tx, max_sat_weight);
+            replacement_vsize = unsigned_tx_max_vbytes(&rbf_psbt.psbt.unsigned_tx, max_sat_weight);
 
+            // Make sure it satisfies RBF rule 4.
             if rbf_psbt
                 .psbt
                 .fee()
                 .expect("has already been sanity checked")
-                >= mempool_entry.fees.descendant + bitcoin::Amount::from_sat(rbf_vsize)
+                >= mempool_entry.fees.descendant + bitcoin::Amount::from_sat(replacement_vsize)
             {
                 return Ok(rbf_psbt);
             }
         }
+
+        unreachable!("We keep increasing the min fee until we run out of funds or satisfy rule 4.")
     }
 
     /// Trigger a rescan of the block chain for transactions involving our main descriptor between
```

Reply:
Do you mean there's an infinite loop in the case that `rbf_vsize` stays the same or always? In some of my tests, there was more than one iteration and, printing the value of `rbf_vsize`, I saw that it increased from one iteration to the next.

I might have overlooked something, but my thinking was that
```
rbf_fee_{i+1} >= minfee_{i+1} = descendant_fee + rbf_vsize_{i}
```
so that if `rbf_vsize_{i} >= rbf_vsize_{i+1}` then 
```
rbf_fee_{i+1} >= descendant_fee + rbf_vsize_{i+1}
```
and so we would exit the loop. So we only continue in the loop if `rbf_vsize` strictly increases.

Reply:
Yes, i'm not quite convinced `rbf_size` would always increase on every iteration.

Reply:
My reasoning above uses the `min_fee` parameter in `create_spend_internal` to ensure the replacement pays at least this fee. and should apply in any case that `rbf_vsize_{i} >= rbf_vsize_{i+1}`. As long as the `min_fee` parameter works as expected, I don't think there can be an infinite loop here.

But I may have overlooked something, so will make the change you suggested to be on the safe side.

Reply:
I've now updated this using your suggestion.